File: agents/livekit/lib/realtime-tts.ts
/**
 * External TTS on a realtime model: the text-output rule (SDK-free).
 *
 * Section 4.3 of docs/livekit-agent-architecture.md: on a realtime
 * (speech-to-speech) model, `options.tts.vendor` set to a vendor other than the
 * model's own provider means the model runs in text-output mode and a discrete
 * TTS stage speaks its text. `options.tts.voice` and `options.tts.language` then
 * belong to that TTS.
 *
 * Same table as agents/pipecat/pipecat_aplisay/realtime_tts.py and the
 * `externalTts` row flag in livekit-model-registry.ts. Keep them in step.
 */
import type { Agent } from "./api-client.js";
import { isLivekitPipelineModelId } from "./livekit-model-registry.js";
import { livekitModelIdFromName } from "./voice-mode.js";

/**
 * The vendor name that means "the model's own voice" for each `livekit:<plugin>/`
 * segment. `google` is both a realtime provider and a TTS vendor, so on a Gemini
 * row it is native and everywhere else it is external.
 */
export const REALTIME_NATIVE_TTS_VENDORS: Record<string, string> = {
  ultravox: "ultravox",
  openai: "openai",
  google: "google",
};

/**
 * Plugins this worker can run in text-output mode. Gemini Live is absent on
 * purpose: no Live model the API still serves accepts a TEXT response modality
 * (checked 2026-09-10; the half-cascade models are gone and the native-audio
 * model rejects it). Must match the rows flagged `externalTts` in
 * livekit-model-registry.ts.
 */
export const TEXT_OUTPUT_PLUGINS: ReadonlySet<string> = new Set(["ultravox", "openai"]);

/** Plugin segment of a `livekit:<plugin>/<model>` name, lowercased. */
export function realtimePlugin(modelName: string): string {
  const m = /^livekit:([^/]+)\//.exec(modelName || "");
  return m ? m[1]!.toLowerCase() : "";
}

/**
 * `options.tts.vendor` normalised for comparison: lowercased, `/model` scoping
 * stripped (`elevenlabs/eleven_flash_v2_5` -> `elevenlabs`). Undefined when unset
 * or blank.
 */
export function ttsVendor(agent: Agent | null | undefined): string | undefined {
  const raw = agent?.options?.tts?.vendor;
  if (typeof raw !== "string") return undefined;
  const vendor = raw.trim().split("/")[0]!.trim().toLowerCase();
  return vendor || undefined;
}

/**
 * The external TTS vendor the agent asks for, or undefined when the agent should
 * use the model's own voice (vendor unset, or the provider's own). This is the
 * rule only; whether the plugin can honour it is {@link textOutputEnabled}.
 */
export function externalTtsVendor(agent: Agent | null | undefined, modelName: string): string | undefined {
  const vendor = ttsVendor(agent);
  if (!vendor) return undefined;
  const plugin = realtimePlugin(modelName);
  const native = REALTIME_NATIVE_TTS_VENDORS[plugin] ?? plugin;
  return vendor === native ? undefined : vendor;
}

/**
 * True when this session must run the realtime model in text-output mode with an
 * external TTS: the agent asks for one AND the plugin supports it on this worker.
 */
export function textOutputEnabled(agent: Agent | null | undefined, modelName: string): boolean {
  // A pipeline row's TTS is always a discrete stage; the rule is for realtime rows.
  if (isLivekitPipelineModelId(livekitModelIdFromName(modelName))) return false;
  if (!TEXT_OUTPUT_PLUGINS.has(realtimePlugin(modelName))) return false;
  return externalTtsVendor(agent, modelName) !== undefined;
}

File: agents/pipecat/pipecat_aplisay/realtime_tts.py
"""External TTS on a realtime model: the text-output rule.

Section 4.3 of docs/livekit-agent-architecture.md: on a realtime
(speech-to-speech) model, ``options.tts.vendor`` set to a vendor other than
the model's own provider means the model runs in text-output mode and a
discrete TTS stage speaks its text. STT and LLM stay inside the provider; only
the TTS stage is decomposed. ``options.tts.voice`` and ``options.tts.language``
then belong to that TTS.

The same table lives server-side in lib/model-voices.js (validation and the
voices API) and as the ``externalTts`` row flag in lib/models/pipecat.js. Keep
the three in step.
"""

from __future__ import annotations

from typing import Optional

#: The vendor name that means "the model's own voice" for each realtime
#: provider segment of a model id (``ultravox/ultravox-v0.7`` -> ``ultravox``).
#: ``google`` is both a realtime provider (Gemini Live) and a TTS vendor, so on
#: a Gemini model it counts as native and on every other model as external.
REALTIME_NATIVE_TTS_VENDORS: dict[str, str] = {
    "ultravox": "ultravox",
    "openai": "openai",
    "google": "google",
}

#: Realtime providers this worker can run in text-output mode. Phase 1 is
#: Ultravox only; OpenAI Realtime and Gemini Live follow (see the plan in the
#: strategy repo). Must match the rows flagged ``externalTts`` in
#: lib/models/pipecat.js.
TEXT_OUTPUT_PROVIDERS: frozenset[str] = frozenset({"ultravox"})


def realtime_provider(model_id: str) -> str:
    """Provider segment of a bare model id (``openai/gpt-realtime`` -> ``openai``)."""
    return (model_id or "").split("/", 1)[0].strip().lower()


def tts_vendor(agent: dict) -> Optional[str]:
    """``options.tts.vendor`` normalised for comparison: lowercased, and with any
    ``/model`` scoping stripped (``elevenlabs/eleven_flash_v2_5`` ->
    ``elevenlabs``). ``None`` when unset or blank."""
    options = (agent or {}).get("options") or {}
    raw = (options.get("tts") or {}).get("vendor")
    if not isinstance(raw, str):
        return None
    vendor = raw.strip().split("/", 1)[0].strip().lower()
    return vendor or None


def external_tts_vendor(agent: dict, model_id: str) -> Optional[str]:
    """The external TTS vendor the agent asks for, or ``None`` when the agent
    should use the model's own voice.

    ``None`` when ``options.tts.vendor`` is unset, or names the provider's own
    vendor (:data:`REALTIME_NATIVE_TTS_VENDORS`). Otherwise the normalised
    vendor name. This is the rule only; whether the provider can honour it is
    :func:`text_output_enabled`.
    """
    vendor = tts_vendor(agent)
    if vendor is None:
        return None
    provider = realtime_provider(model_id)
    native = REALTIME_NATIVE_TTS_VENDORS.get(provider, provider)
    if vendor == native:
        return None
    return vendor


def text_output_enabled(agent: dict, model_id: str) -> bool:
    """True when this session must run the realtime model in text-output mode
    with an external TTS: the agent asks for one AND the provider supports it
    on this worker."""
    if realtime_provider(model_id) not in TEXT_OUTPUT_PROVIDERS:
        return False
    return external_tts_vendor(agent, model_id) is not None
